### ISSUE TITLE
Weired fine-detailed reconstruction for can class

### ISSUE THREAD

**Issue by HanzhiC** [zju3dv/gcasp#5]
Hey I was testing the reconstruction quality of the byproduct reconstruction. The meshes of the other objects look good to me except for the **Can class**, which always gives out tiny shapes.

I think the pose and scale are correctly applied to transform mesh from canonical space to camera space, as the other objects look very reasonable. I suspect there are some problems with the pre-trained DualSDF-weight, as I observe that during shape optimization in the `optimize_shape_ransac` function, the decoded shape from the feature is always tiny.

could you kindly see what is going on with the mesh reconstruction? 

Here are some examples:
![1](https://user-images.githubusercontent.com/47291733/213790210-0db63c91-92c5-4a29-b747-66fb23a103c6.png)
![2](https://user-images.githubusercontent.com/47291733/213790216-d0dc7edf-5cfb-4183-a79a-26ca859ffdf1.png)

Thank you!

**Comment by liguanglin** [comment]
Hi, actually we met this problen before. 

We think this is because the number of CAD models (can category) used to train DualSDF is small. This makes the fine shape not consistent with the optimized coarse shape. 

To show the mesh reconstruction in the supplementary material, we made data augmentation on given CAD models (likes stretching the cans) to make the DualSDF more stable. To evaluate the reconstructed shapes, you could use the coarse shapes, or try the above method to make DualSDF stable.

**Comment by HanzhiC** [comment]
> Hi, actually we met this problem before.
> 
> We think this is because the number of CAD models (can category) used to train DualSDF is small. This makes the fine shape not consistent with the optimized coarse shape.
> 
> To show the mesh reconstruction in the supplementary material, we made data augmentation on given CAD models (like stretching the cans) to make the DualSDF more stable. To evaluate the reconstructed shapes, you could use the coarse shapes, or try the above method to make DualSDF stable.

Could you elaborate on what you mean by "using coarse shapes"?

**Comment by liguanglin** [comment]
> Could you elaborate on what you mean by "using coarse shapes" (low resolution for marching cube, or use prim_sdf_fun as rendering function, or?)? Also, is this data augmentation used in test or train time?

Coarse shape means semantic primitives, the spheres generated by [prim_sdf_fun](https://github.com/zju3dv/gcasp/blob/main/DualSDFHandler.py#L180). You could treat the spheres' centers as the reconstructed point cloud, and measure the chamfer distance between the reconstructed point cloud and the ground truth shape as shown in the paper. 

The above data augmentation is used to expand the training samples of CAD models to pre-train a DualSDF model  of the can category.

**Comment by HanzhiC** [comment]
Thanks for such a quick response. Now see the point.
One last question, is the weight pre-trained with such a data augmentation strategy available?

**Comment by liguanglin** [comment]
Yes, but cannot provide in time due to lab closure during spring festival.

**Comment by HanzhiC** [comment]
Take your time. It would be wonderful if you could provide it some time. 
Thank you.

**Comment by liguanglin** [comment]
Hi, you could try to replace the weights of **can category** with [epoch_9999.pth](https://drive.google.com/file/d/1tHfe-Ql4fiIJPo-anCp00oJf6RqcULqm/view?usp=sharing) and replace the segmetation weights with [epoch_14_iters_290640.pth](https://drive.google.com/file/d/1bsg1G5SvZJnKD_Y-xz7BIEBGDdQVxnyp/view?usp=sharing).
